pkg/util/grpcutil/grpc_util.go

```diff
@@ -105,6 +105,17 @@ func IsClosedConnection(err error) bool {
 	return netutil.IsClosedConnection(err)
 }
 
+// IsDecommissionedError returns true if err's Cause is an error produced by gRPC
+// due to remote node has been decommissioned.
+func IsDecommissionedError(err error) bool {
+	if s, ok := status.FromError(errors.UnwrapAll(err)); ok {
+		if s.Code() == codes.FailedPrecondition && strings.Contains(err.Error(), "was permanently removed from the cluster") {
```

Comment:
String matching is not OK here. We need to convince ourselfes that at the caller, `FailedPrecondition` uniquely communicates a decommissioned peer in the place in which we're getting the error. You're interested in the `OnOutgoingPing` interceptor, right? Luckily this is called directly (i.e. not underneath gRPC) so the structure of the error remains but even better, we can adjust the signature, so I think you can change the signature of `OnOutgoingPing` to something like `(decommissioned bool, err error)`:

```go
		OnOutgoingPing: func(ctx context.Context, req *rpc.PingRequest) (decommissioned bool, _ error) {
			// Outgoing ping will block requests with codes.FailedPrecondition to
			// notify caller that this replica is decommissioned but others could
			// still be tried as caller node is valid, but not the destination.
			err := checkPingFor(ctx, req.TargetNodeID, codes.FailedPrecondition)
			return err != nil, err
		},
```

Then you shouldn't need to add `IsDecommissionedError` here.

It would be nice to have a "globally unique" way to determine whether an error is about a decommissioned node, but last time I checked there was a lot of protobuf marshalling related madness due to `*grpc.Status`; let's not stir that pot now.

Reply:
Ah, in case it's not clear, there wouldn't be a method `IsDecommissionedError` in this package at the end of this.